Use dedicated class for representing Events

stopfstedt opened the pull request:
this allows us to separate derived values, such as `calendarStartYear` and `calendarEndYear`, from raw data points, such as `startDate` and `endDate`, using cached getters.

the sets us up for a cleaner representation going forward.

refs #8956

stopfstedt committed:
create dedicated Event class and refactor derived attrs as getters.

stopfstedt committed:
streamline event generation by replacing input pre-processing with input processing in constructor.

all the data wrangling happens directly in the constructor now.
this eliminates the need to a factory in the user/school events service.
furthermore, Event.slug is now be cached getter as it should be - this required some re-shuffling of utility methods.

this required a decent amount of rewiring in integration test coverage where we
have to now use actual Event objects as component input instead of mock
object that were hand-rolled or provided by mirage.

jrjohnson commented on packages/test-app/tests/classes/event-test.js:
```diff
@@ -0,0 +1,118 @@
+
+module('Unit | Classes | Event', function (hooks) {
+  setupTest(hooks);
+  test.each(
```
Didn't know about this!

stopfstedt replied:
it's great, innit?

stopfstedt committed:
makes private methods of the module private methods of the class.

this is more straightforward.

stopfstedt committed:
adds more test coverage for Event class.

the Event class does a data transformation on its input in the constructor,
this covers the transformed attributes.

stopfstedt commented on packages/ilios-common/addon/classes/event.js:
```diff
@@ -0,0 +1,153 @@
+ * @param { Number|undefined } ilmSessionId The event's ILM session ID
+ * @return { String }
+ */
+function getSlugForUserEvent(startDate, offeringId, ilmSessionId) {
```
handled :heavy_check_mark:

jrjohnson reviewed (approved):
💟

dartajax reviewed (approved):
all calendars checked - looks good to me
